### ISSUE TITLE
GLM-10B model inference is slower than PT2.0+CUDA in BF16 on GPU

### ISSUE THREAD

**Issue by wbmc** [pytorch/xla#5116]
## ❓ Questions and Help
Recently, I speed up [GLM-10B](https://huggingface.co/THUDM/glm-10b-chinese/blob/main/modeling_glm.py) model inference with `torch/xla`.
The XLA got **3.5x** faster than PT2.0+CUDA in **FP32** on **GPU(A100)**.  (cuda:280ms ->  xla:80ms)
But It was **30%** slower than PT2.0+CUDA when I change to **BF16** mode (by set XLA_USE_BF16=1). (cuda:39ms ->
 xla:54ms)

What could be the cause of this problem？
Thanks! 
cc @JackCaoG

# Reproduce
## ENVIRONMENT
1 x A100 GPU 80GB
PT2.0 + CUDA
PT2.0 + XLA

## STEP 1
`download the following archive packs into a directory and extract files from them`
[code.zip](https://github.com/pytorch/xla/files/11643270/code.zip)  
[debug_info_bf16.tar.gz](https://github.com/pytorch/xla/files/11643239/debug_info_bf16.tar.gz)
## STEP2
Install requirements:
`pip install transformers huggingface_hub sentencepiece`
## STEP3
run `CUDA_VISIBLE_DEVICES=0 PJRT_DEVICE=GPU GPU_NUM_DEVICES=1 python test.py`

# METRICS 
[MetricsData; step=21]
Metric: DeviceLockWait
  TotalSamples: 44
  Accumulator: 264.446us
  ValueRate: 018.323us / second
  Rate: 3.04868 / second
  Percentiles: 1%=001.051us; 5%=001.117us; 10%=001.162us; 20%=001.229us; 50%=004.821us; 80%=006.350us; 90%=007.285us; 95%=009.382us; 99%=091.206us
Metric: InputOutputAliasCount
  TotalSamples: 2
  Accumulator: 299.00
  ValueRate: 39.05 / second
  Rate: 0.261229 / second
  Percentiles: 1%=3.00; 5%=3.00; 10%=3.00; 20%=3.00; 50%=296.00; 80%=296.00; 90%=296.00; 95%=296.00; 99%=296.00
Metric: IrValueTensorToXlaData
  TotalSamples: 359
  Accumulator: 867ms763.890us
  ValueRate: 039ms178.479us / second
  Rate: 16.2271 / second
  Percentiles: 1%=016.087us; 5%=018.445us; 10%=019.246us; 20%=021.188us; 50%=030.229us; 80%=003ms279.683us; 90%=004ms352.670us; 95%=004ms363.898us; 99%=005ms425.543us
Metric: TensorToData
  TotalSamples: 361
  Accumulator: 866ms322.604us
  ValueRate: 039ms158.520us / second
  Rate: 16.3175 / second
  Percentiles: 1%=015.085us; 5%=017.429us; 10%=018.243us; 20%=020.072us; 50%=029.038us; 80%=003ms262.017us; 90%=004ms350.673us; 95%=004ms362.207us; 99%=005ms424.136us
Metric: TensorsGraphSize
  TotalSamples: 22
  Accumulator: 69652.00
  ValueRate: 4825.93 / second
  Rate: 1.5243 / second
  Percentiles: 1%=3166.00; 5%=3166.00; 10%=3166.00; 20%=3166.00; 50%=3166.00; 80%=3166.00; 90%=3166.00; 95%=3166.00; 99%=3166.00
Metric: UnwrapXlaData
  TotalSamples: 5567
  Accumulator: 953.291us
  ValueRate: 070.414us / second
  Rate: 361.887 / second
  Percentiles: 1%=000.052us; 5%=000.056us; 10%=000.060us; 20%=000.067us; 50%=000.105us; 80%=000.129us; 90%=000.151us; 95%=000.179us; 99%=000.285us
Metric: WrapXlaData
  TotalSamples: 19001
  Accumulator: 014ms253.076us
  ValueRate: 002ms748.168us / second
  Rate: 2525.19 / second
  Percentiles: 1%=000.344us; 5%=000.361us; 10%=000.371us; 20%=000.388us; 50%=000.423us; 80%=000.484us; 90%=000.525us; 95%=000.573us; 99%=000.752us
Counter: CachedCompile
  Value: 20
Counter: CreateXlaTensor
  Value: 40333
Counter: DestroyLtcTensor
  Value: 39207
Counter: DestroyXlaTensor
  Value: 39207
Counter: DeviceDataCacheMiss
  Value: 2
Counter: RegisterXLAFunctions
  Value: 1
Counter: UncachedCompile
  Value: 2
Counter: xla::_copy_from
  Value: 2999
Counter: xla::_softmax
  Value: 528
Counter: xla::_unsafe_view
  Value: 3190
Counter: xla::add
  Value: 3740
Counter: xla::addcmul
  Value: 1078
Counter: xla::arange_out
  Value: 22
Counter: xla::bmm
  Value: 1056
Counter: xla::div
  Value: 528
Counter: xla::embedding_symint
  Value: 66
Counter: xla::empty_symint
  Value: 931
Counter: xla::expand_symint
  Value: 2112
Counter: xla::gelu
  Value: 528
Counter: xla::index_select
  Value: 66
Counter: xla::mm
  Value: 2134
Counter: xla::mul
  Value: 1056
Counter: xla::native_batch_norm
  Value: 1078
Counter: xla::permute
  Value: 2112
Counter: xla::rsub
  Value: 528
Counter: xla::select
  Value: 44
Counter: xla::slice
  Value: 1232
Counter: xla::split
  Value: 528
Counter: xla::stack
  Value: 22
Counter: xla::t
  Value: 2134
Counter: xla::transpose
  Value: 528
Counter: xla::unsqueeze
  Value: 66
Counter: xla::view_symint
  Value: 13992
Counter: xla::zero_
  Value: 22
Metric: CompileTime
  TotalSamples: 2
  Accumulator: 13s976ms706.000us
  ValueRate: 02s073ms829.996us / second
  Rate: 0.319494 / second
  Percentiles: 1%=06s791ms878.616us; 5%=06s791ms878.616us; 10%=06s791ms878.616us; 20%=06s791ms878.616us; 50%=07s185ms827.384us; 80%=07s185ms827.384us; 90%=07s185ms827.384us; 95%=07s185ms827.384us; 99%=07s185ms827.384us
Metric: ExecuteTime
  TotalSamples: 21
  Accumulator: 01s308ms776.164us
  ValueRate: 093ms325.137us / second
  Rate: 1.4986 / second
  Percentiles: 1%=061ms143.041us; 5%=061ms237.601us; 10%=061ms267.610us; 20%=061ms301.001us; 50%=061ms425.266us; 80%=061ms474.416us; 90%=062ms551.071us; 95%=071ms595.959us; 99%=071ms802.000us
Metric: OutboundData
  TotalSamples: 361
  Accumulator: 680.52MB
  ValueRate: 30.76MB / second
  Rate: 16.3175 / second
  Percentiles: 1%=256.00B; 5%=256.00B; 10%=2.00KB; 20%=2.00KB; 50%=6.00KB; 80%=6.00MB; 90%=8.00MB; 95%=8.00MB; 99%=8.00MB
Metric: TransferToServerTime
  TotalSamples: 361
  Accumulator: 865ms741.611us
  ValueRate: 039ms086.979us / second
  Rate: 16.3175 / second
  Percentiles: 1%=012.823us; 5%=014.676us; 10%=015.413us; 20%=017.095us; 50%=024.513us; 80%=003ms257.652us; 90%=004ms346.752us; 95%=004ms357.965us; 99%=005ms419.138us
Counter: CreateCompileHandles
  Value: 2
Counter: CreateDataHandles
  Value: 18147
Counter: MarkStep
  Value: 22

**Comment by JackCaoG** [comment]
Can you share the HLO? You can do that by `XLA_SAVE_TENSORS_FMT='hlo' XLA_SAVE_TENSORS_FILE='/tmp/save1.hlo'` and rerun the model.

**Comment by wbmc** [comment]
> Can you share the HLO? You can do that by `XLA_SAVE_TENSORS_FMT='hlo' XLA_SAVE_TENSORS_FILE='/tmp/save1.hlo'` and rerun the model.

Yes, This is the model hlo file. Thank you!
cc @JackCaoG 
[save_bf16_hlo.zip](https://github.com/pytorch/xla/files/11659582/save_bf16_hlo.zip)

**Comment by JackCaoG** [comment]
Thanks, I opened an issue to xla:GPU team. Are you comparing XLA:GPU with CUDA or indcutor?

**Comment by wbmc** [comment]
> Thanks, I opened an issue to xla:GPU team. Are you comparing XLA:GPU with CUDA or indcutor?

Thanks,
Do you mean that comparing TensorFlow+XLA:GPU with PyTorch+CUDA? Unfortunately, GLM-10B model doesn't has a TF version.
I compared the PyTorch+XLA::GPU with PyTorch+CUDA. The XLA:GPU was 30% slower than CUDA version on BF16

**Comment by JackCaoG** [comment]
oh, I mean were you comparing pytorch+XLA:GPU with CUDA or inductor+Dynamo. Seems like you compared it with the CUDA(pytorch eager mode) directly.

**Comment by wbmc** [comment]
> oh, I mean were you comparing pytorch+XLA:GPU with CUDA or inductor+Dynamo. Seems like you compared it with the CUDA(pytorch eager mode) directly.
Yes, I compared it with the CUDA (pytorch eager mode)

**Comment by wbmc** [comment]
> > oh, I mean were you comparing pytorch+XLA:GPU with CUDA or inductor+Dynamo. Seems like you compared it with the CUDA(pytorch eager mode) directly.
> 
> Yes, I compared it with the CUDA (pytorch eager mode)

**Comment by kevint324** [comment]
@wbmc 
how do you get rid of the recompilations from dynamic shapes?

**Comment by wbmc** [comment]
> @wbmc how do you get rid of the recompilations from dynamic shapes?

Padding

**Comment by kevint324** [comment]
@wbmc  could you please share some code on how to do the padding
